### DISCUSSION TITLE
[HttpClient] RetryableHttpClient sends to RetryStrategyInterface::getDelay always empty $responseContent

### DISCUSSION THREAD

**Discussion by Tetragramat** [symfony/symfony#53723] [Q&A] [answered]
### Symfony version(s) affected

5.4.31

### Description

When using `RetryableHttpClient` together with custom retry strategy `RetryStrategyInterface` then when I receive for example 429 with response content then that content is never passed into `RetryStrategyInterface::getDelay` `$responseContent`.

That means when I get response with http code 429 and response content "The rate limit for this resource is 300/minute. Please try again in 35 seconds" then I'm unable to make custom retry strategy that parses response content and output delay of 35 seconds. If I try to read `$responseContent` then it's empty. If I try read `$context->getResponse()->getContent()` then response is cancelled.

### How to reproduce

I don't know how to test it easier. 
Make `RetryableHttpClient` request to action that returns `new Response('The rate limit for this resource is 300/minute. Please try again in 35 seconds', 429);`. (do NOT add `retry-after` header)
In `\Symfony\Component\HttpClient\Retry\GenericRetryStrategy::getDelay` dump `$responseContent` to see if it's empty or not.

**Comment by nicolas-grekas** [comment, marked as answer]
IIRC, you need to make your strategy return null when shouldRetry is called with a null responseContent.
Then, getDelay will receive the content as expected.

**Reply by Tetragramat** [reply]
I've tried it and works. Thanks.

There is just a little problem with documentation of this behaviour. I see it documented at `\Symfony\Component\HttpClient\Retry\RetryStrategyInterface::shouldRetry` as `@return bool|null Returns null to signal that the body is required to take a decision` but when I look only at getDelay since I'm just extending `\Symfony\Component\HttpClient\Retry\GenericRetryStrategy` then I won't notice that `$responseContent` will be null unless certain conditions are met.
So I believe that `\Symfony\Component\HttpClient\Retry\RetryStrategyInterface::getDelay` should get mention of that behaviour too.

Example phpdoc to notify user about `$responseContent` unexpected behaviour:
```php
    /**
     * Returns the time to wait in milliseconds.
     *
     * @param ?string $responseContent Always null unless {@see shouldRetry} returns null when its $responseContent is null
     */
    public function getDelay(AsyncContext $context, ?string $responseContent, ?TransportExceptionInterface $exception): int;
```

**Reply by nicolas-grekas** [reply]
Nice. Doc PR welcome of course!
